src/main/java/dev/Roach/DealsFetcher.java

```diff
@@ -0,0 +1,115 @@
+package dev.Roach;
+
+import java.io.FileWriter;
+import java.io.IOException;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.util.ArrayList;
+import java.util.List;
+
+
+public class DealsFetcher {
+
+    final HttpClient client;
+
+    public DealsFetcher(HttpClient client) {
+        this.client = client;
+    }
+
+    public DealsFetcher() {
+        this.client = HttpClient.newBuilder().build();
+    }
+
+    //My solution, around 1min 30s
+    //I added writing to the file to omit api calls limit.
+    public List<String> getAllDeals() {
+
+        List<String> allPages = new ArrayList<>();
+
+        try (FileWriter fw = new FileWriter("listOfAllDeals.txt")){
+            int pageNumber = 0;
+            int totalPages = Integer.MAX_VALUE;
```

Comment:
Maybe let's poll for first 10 pages, additionally, please experiment with increasing page size to decrease the number of calls.